### PR TITLE
ZUBA-245 Rerank TopK References

### PR DESCRIPTION
**Pull request by dbarkowsky**
## 🎯 Summary

[ZUBA-245](https://citz-imb.atlassian.net/browse/ZUBA-245?atlOrigin=eyJpIjoiYjA2MTdhYjliMjBkNGVjNzk2YjZhNGVjN2ZjNjliZGYiLCJwIjoiaiJ9)

## Changes

- Added env to the docker compose. Can now put a .env in the .docker folder. Avoid committing keys. Removed the hardcoded bits from entrypoint.sh as part of this. Would be good to have a file documenting which variables are needed.
- Added some things to the .gitignore. Had a lot of downloaded files from the examples folder that were getting tracked by git.
- aiohttp==3.10.5 added to requirements file
- Re-rank code added to rag.py. Can take a list of document fields to do tiered sorting.
- Updated trulens App_Id to track these results separately.
- Small readme update. There's a lot more to do here in the future, especially with the examples README.

Apologies for the auto-formatting my IDE did. Maybe we can consider a formatting standard.

## 🔰 Checklist

- [x] I have read and agree with the following checklist

> - I have performed a self-review of my code.
> - I have commented my code, particularly in hard-to-understand areas.
> - I have made corresponding changes to the documentation where required.
> - I have tested my changes to the best of my ability.
> - I have consulted with the team if introducing a new dependency.

### CODE REVIEW

File: web/backend-fastapi/app/models/rag.py [hunk 1]
```diff
@@ -65,29 +74,52 @@ def create_prompt(self, query: str, context_str: str, chat_history: List[ChatHis
...
+
     @instrument
     def query(self, query: str, chat_history: List[ChatHistory], embeddings, kg) -> str:
         context_str = self.retrieve(query, embeddings, kg)
+        # Rerank once to order for LLM
+        context_str = self.re_rank_reference(context_str, query)
         create_prompt = self.create_prompt(query, context_str, chat_history)
         bedrock_response = self.get_response(create_prompt)
+        # Rerank again to sort references by relevance to response
+        context_str = self.re_rank_reference(
```
Comment (antsand) [inline web/backend-fastapi/app/models/rag.py:119]: Why do we need a double rerank? Is the performance better with dobule rerank? seems strange.

File: .docker/compose.controller.yaml [hunk 2]
```diff
@@ -260,6 +260,8 @@ services:
       NEO4J_VECTOR_INDEX: ${NEO4J_VECTOR_INDEX:-Acts_Updatedchunks}
       NEO4J_DB: ${NEO4J_DB:-neo4j}
       NEO4J_PORT: ${NEO4J_PORT:-7687}
+      AWS_ACCESS_KEY_ID: ${AWS_ACCESS_KEY_ID}
```
Comment (antsand) [inline .docker/compose.controller.yaml:263]: one of the issues we faced with this was sometimes fast api is loaded before the environment variables gets loaded. So we added to the entrypoint.sh to make sure fastapi cannot start till the variables are loaded.  We can keep it here.

File: web/backend-fastapi/entrypoint.sh [hunk 3]
```diff
@@ -3,8 +3,6 @@
 # Load environment variables
 echo "Loading environment variables..."
 
-export AWS_ACCESS_KEY_ID="--"
```
Comment (antsand) [inline web/backend-fastapi/entrypoint.sh:6]: As mentioned above. We may need to prune and test this to be certain we don't need this. I believe we had timing issues (fastapi starting without the variables loaded) that we created this file. I'd be cautious removing this before thorough testing. For now, my 2cents is to keep it.  

This is local so.. just an inconvenience more than anything else.

File: web/backend-fastapi/app/models/rag.py (hunk 1 above)
Comment (dbarkowsky) [reply web/backend-fastapi/app/models/rag.py:119]: I would be good to remove the initial rerank prior to the bedrock request. I'm not convinced that it improves bedrock's response, but that would mean not reranking the references according to the user's query, as suggested in standup.

If you mean why rerank on two different fields, it seemed more consistent to prioritize which Act the information came from then order the references based on their text relevance. Just doing it by text didn't always land the expected Act at the top of the list.

File: web/backend-fastapi/entrypoint.sh (hunk 3 above)
Comment (dbarkowsky) [reply web/backend-fastapi/entrypoint.sh:6]: Is there a way we can test this? It seems like an accident waiting to happen with someone potentially committing a key.

File: .docker/compose.controller.yaml (hunk 2 above)
Comment (dbarkowsky) [reply .docker/compose.controller.yaml:263]: The entrypoint.sh script still checks if the access key variable is populated, so it should still block fastapi from starting.

**Review by antsand — APPROVED** [review]

**Pull request merged**